src/wayland/socket.rs

```diff
@@ -0,0 +1,146 @@
+//! let listening_socket = ListeningSocketSource::new_auto(None).unwrap();
+//!
+//! event_loop.handle().insert_source(listening_socket, |client_stream, _, state| {
+//!     // Inside the callback, you should insert the client into the display.
+//!     //
+//!     // You may also associate some data with the client when inserting the client.
+//!     state.display.insert_client(client_stream, Arc::new(ExampleClientData));
+//! });
+//!
+//! # struct ExampleClientData;
+//! #
+//! # impl<D> wayland_server::backend::ClientData<D> for ExampleClientData {
+//! #     fn initialized(&self, _: wayland_server::backend::ClientId) {}
+//! #     fn disconnected(
+//! #         &self,
+//! #         _: wayland_server::backend::ClientId,
+//! #         _: wayland_server::backend::DisconnectReason
+//! #     ) {}
+//! # }
+//! ```
+
+use std::{ffi::OsStr, io, os::unix::net::UnixStream};
+
+use calloop::{
+    generic::Generic, EventSource, Interest, Mode, Poll, PostAction, Readiness, Token, TokenFactory,
+};
+use wayland_server::socket::{BindError, ListeningSocket};
+
+/// A Wayland listening socket event source.
+///
+/// This implements [`EventSource`] and may be inserted into an event loop.
+#[derive(Debug)]
+pub struct ListeningSocketSource {
+    logger: slog::Logger,
+    socket: Generic<ListeningSocket>,
+}
+
+impl ListeningSocketSource {
+    /// Creates a new listening socket, automatically choosing the next available `wayland` socket name.
+    pub fn new_auto<L>(logger: L) -> Result<ListeningSocketSource, BindError>
+    where
+        L: Into<Option<::slog::Logger>>,
+    {
+        // Try socket numbers 1-32. Remember the upper bound of Range is exclusive.
+        //
+        // We don't try wayland-0 due since clients may connect to the wrong compositor. Clients these days
+        // should be connecting based off the WAYLAND_DISPLAY or WAYLAND_SOCKET environment variables.
+        let socket = ListeningSocket::bind_auto("wayland", 1..33)?;
+
+        let logger = crate::slog_or_fallback(logger)
+            .new(slog::o!("wayland_socket" => format!("{}", socket.socket_name().to_string_lossy())));
+        slog::info!(logger, "Created new socket");
+
+        Ok(ListeningSocketSource {
+            logger,
+            socket: Generic::new(socket, Interest::BOTH, Mode::Level),
+        })
+    }
+
+    /// Creates a new listening socket with the specified name.
+    pub fn with_name<L>(name: &str, logger: L) -> Result<ListeningSocketSource, BindError>
+    where
+        L: Into<Option<::slog::Logger>>,
+    {
+        let socket = ListeningSocket::bind(name)?;
+        let logger = crate::slog_or_fallback(logger)
+            .new(slog::o!("wayland_socket" => format!("{}", socket.socket_name().to_string_lossy())));
+        slog::info!(logger, "Created new socket");
+
+        Ok(ListeningSocketSource {
+            logger,
+            socket: Generic::new(socket, Interest::BOTH, Mode::Level),
```

Comment:
Just to make sure that this is intentional, before 0.30 in anvil we used to do `Interest::READ`, is the change to BOTH intentional?

Reply:
Probably a mistake, let me look into it further.